File: Embedded/Friends.cpp
#include "Friends.h"
#include "Requests.h"

#define BACKGROUND TFT_BLACK
#define FONTSIZE 1

Message_Module::Message_Module(TFT_eSPI* tft_to_use, MPU9255* imu_to_use, char* username):type_module(message_to_type,imu_to_use){
  state=0;
  is_active=false;
  tft=tft_to_use;
  imu=imu_to_use;
  number_of_messages=0;
  start_of_messages=0;
  user=username;
  check_message_timer=millis()-10000;
  for(int i=0;i<10;i++)
  {
    messages_buffer[i]=new char[200];
  }
}

void Message_Module::activate(){
  type_module.reset();
  is_active=true;
  tft->fillRect(0,20,130,140,BACKGROUND);
  if(number_of_messages!=0) {
    tft->setCursor(0,20,FONTSIZE);
    char to_print[100];
    sprintf(to_print,"You have %d new messages",number_of_messages);
    tft->println(to_print);
    tft->println(messages_buffer[start_of_messages]);
  } else {
    tft->setCursor(0,20,FONTSIZE);
    tft->println("You have no new message");
    tft->print("Long Press and input the username of your friend");
  }
}

void Message_Module::close() {
  state=0;
  is_active=false;
}

void Message_Module::update(int bv) {
  if(is_active) {
    if(number_of_messages!=0) {
      if(bv==1) {
        number_of_messages--;
        start_of_messages=(start_of_messages+1)%10;
        tft->fillRect(0,20,130,140,BACKGROUND);
        if(number_of_messages!=0) {
          tft->setCursor(0,20,FONTSIZE);
          char to_print[100];
          sprintf(to_print,"You have %d new message(s)",number_of_messages);
          tft->println(to_print);
          tft->println(messages_buffer[start_of_messages]);
        } else {
          tft->setCursor(0,20,FONTSIZE);
          tft->println("You have no new message");
          tft->print("Long Press and input the username of your friend");
        }
      }
    }
    else {
      if(state==0) {
        if(type_module.update(bv)==1) {
          state=1;
          tft->fillRect(0,20,130,140,BACKGROUND);
          tft->setCursor(0,20,FONTSIZE);
          tft->print(message_to_type);
        }
      }
      else if(state==1) {
        int a=type_module.update(bv);
        if(a==1) {
          tft->fillRect(0,20,130,140,BACKGROUND);
          tft->setCursor(0,20,FONTSIZE);
          tft->print(message_to_type);
        }
        if(a==2) {
          state=2;
          tft->fillRect(0,20,130,140,BACKGROUND);
          tft->setCursor(0,20,FONTSIZE);
          tft->print("Long Press and input the content");
          //Serial.println("I am going to send message to this username:");
          //Serial.println(message_to_type);
          sprintf(body,"module=message&user1=%s&user2=%s",user,message_to_type);
        }
      }
      else if(state==2) {
        if(type_module.update(bv)==1) {
          state=3;
          tft->fillRect(0,20,130,140,BACKGROUND);
          tft->setCursor(0,20,FONTSIZE);
          tft->print(message_to_type);
        }
      }
      else if(state==3) {
        int a=type_module.update(bv);
        if(a==1) {
          tft->fillRect(0,20,130,140,BACKGROUND);
          tft->setCursor(0,20,FONTSIZE);
          tft->print(message_to_type);
        }
        if(a==2) {
          state=4;
          //Serial.println("I am going to send the content:");
          //Serial.println(message_to_type);
          sprintf(body+strlen(body),"&content=%s",message_to_type);
        }
      }
      else if(state==4) {
        tft->fillRect(0,20,130,140,BACKGROUND);
        tft->setCursor(0,20,FONTSIZE);
        tft->print("Please Wait");
        sprintf(request,"POST http://608dev.net/sandbox/sc/haoyangg/final_project/friends.py HTTP/1.1\r\n");
        strcat(request,"Host: 608dev.net\r\n");
        strcat(request,"Content-Type: application/x-www-form-urlencoded\r\n");
        sprintf(request+strlen(request),"Content-Length: %d\r\n", strlen(body)); //append string formatted to end of request buffer
        strcat(request,"\r\n"); //new line from header to body
        strcat(request,body); //body
        strcat(request,"\r\n"); //header
        do_http_request("608dev.net", request, response, 1000, 6000,true);
        tft->fillRect(0,20,130,140,BACKGROUND);
        tft->setCursor(0,20,FONTSIZE);
        if(response[0]!='\0') {
        tft->println(response);
        }
        else {
          tft->println("Failed to connect server! Please check your wifi connection");
        }
        tft->println("Long Press to send a new message");
        state=0;
      }
    }
  }
  else {
    if((millis()-check_message_timer>10000||millis()<check_message_timer)&&(!do_not_request)) {
      sprintf(request,"GET http://608dev.net/sandbox/sc/haoyangg/final_project/friends.py?module=message&user=%s&space=%d HTTP/1.1\r\n",user,10-number_of_messages);
      strcat(request,"Host: 608dev.net\r\n");
      strcat(request,"\r\n");
      do_http_request("608dev.net", request, response, 1000, 6000,true);
      char* p=strtok(response,",");
      while(p!=NULL) {
        sprintf(messages_buffer[(start_of_messages+number_of_messages)%10],p);
        number_of_messages++;
        p=strtok(NULL,",");
      }
      check_message_timer=millis();
    }
  }
}

Add_Module::Add_Module(TFT_eSPI* tft_to_use, MPU9255* imu_to_use,  char* username):type_module(message_to_type,imu_to_use) {
  state=0;
  is_active=false;
  tft=tft_to_use;
  imu=imu_to_use;
  number_of_requests=0;
  start_of_requests=0;
  user=username;
  check_requests_timer=millis()-10000;
  for(int i=0;i<10;i++) {
    requests_buffer[i]=new char[50];
  }
}

void Add_Module::activate() {
  type_module.reset();
  is_active=true;
  tft->fillRect(0,20,130,140,BACKGROUND);
  if(number_of_requests!=0) {
    tft->setCursor(0,20,FONTSIZE);
    char to_print[200];
    sprintf(to_print,requests_buffer[start_of_requests]);
    sprintf(to_print+strlen(to_print)," wants to be your friend.\nShort press to accept.\nLong press to refuse");
    tft->print(to_print);
  }
  else {
    tft->setCursor(0,20,FONTSIZE);
    tft->print("Long Press to add friends");
  }
}

void Add_Module::close() {
  state=0;
  is_active=false;
}
void Add_Module::update(int bv) {
  if(is_active) {
    if(number_of_requests!=0) {
      if(bv==1) {  //On the server side, send your friend a message to let him know.
        sprintf(body,"module=add&user1=%s&user2=%s&action=accept",user,requests_buffer[start_of_requests]);
        sprintf(request,"POST http://608dev.net/sandbox/sc/haoyangg/final_project/friends.py HTTP/1.1\r\n");
        strcat(request,"Host: 608dev.net\r\n");
        strcat(request,"Content-Type: application/x-www-form-urlencoded\r\n");
        sprintf(request+strlen(request),"Content-Length: %d\r\n", strlen(body)); //append string formatted to end of request buffer
        strcat(request,"\r\n"); //new line from header to body
        strcat(request,body); //body
        strcat(request,"\r\n"); //header
        do_http_request("608dev.net", request, response, 1000, 6000,true);
        number_of_requests--;
        start_of_requests=(start_of_requests+1)%10;
        tft->fillRect(0,20,130,140,BACKGROUND);
        if(number_of_requests!=0) {
          tft->setCursor(0,20,FONTSIZE);
          char to_print[200];
          sprintf(to_print,requests_buffer[start_of_requests]);
          sprintf(to_print+strlen(to_print)," wants to be your friend.\nShort press to accept.\nLong press to refuse");
          tft->print(to_print);
        }
        else {
          tft->setCursor(0,20,FONTSIZE);
          tft->println("You have no friends requests");
          tft->print("Long Press to add friends");
        }
      }
      if(bv==2) {
        Serial.println("I will refuse this requests");//On the server side, send your friend a message to let him know.
        sprintf(body,"module=add&user1=%s&user2=%s&action=refuse",user,requests_buffer[start_of_requests]);
        sprintf(request,"POST http://608dev.net/sandbox/sc/haoyangg/final_project/friends.py HTTP/1.1\r\n");
        strcat(request,"Host: 608dev.net\r\n");
        strcat(request,"Content-Type: application/x-www-form-urlencoded\r\n");
        sprintf(request+strlen(request),"Content-Length: %d\r\n", strlen(body)); //append string formatted to end of request buffer
        strcat(request,"\r\n"); //new line from header to body
        strcat(request,body); //body
        strcat(request,"\r\n"); //header
        do_http_request("608dev.net", request, response, 1000, 6000,true);
        number_of_requests--;
        start_of_requests=(start_of_requests+1)%10;
        tft->fillRect(0,20,130,140,BACKGROUND);
        if(number_of_requests!=0) {
          tft->setCursor(0,20,FONTSIZE);
          char to_print[200];
          sprintf(to_print,requests_buffer[start_of_requests]);
          sprintf(to_print+strlen(to_print)," wants to be your friend.\nShort press to accept.\nLong press to refuse");
          tft->print(to_print);
        }
        else {
          tft->setCursor(0,20,FONTSIZE);
          tft->println("You have no friends requests");
          tft->print("Long Press to add friends");
        }
      }
    }
    else {
      if(state==0) {
        if(type_module.update(bv)==1) {
          state=1;
          tft->fillRect(0,20,130,140,BACKGROUND);
          tft->setCursor(0,20,FONTSIZE);
          tft->print(message_to_type);
        }
      }
      else if(state==1) {
        int a=type_module.update(bv);
        if(a==1) {
          tft->fillRect(0,20,130,140,BACKGROUND);
          tft->setCursor(0,20,FONTSIZE);
          tft->print(message_to_type);
        }
        if(a==2) {
          state=2;
          //Serial.println("I am going to add this friend:");
          //Serial.println(message_to_type);
        sprintf(body,"module=add&user1=%s&user2=%s&action=send",user,message_to_type);
        }
      }
      else if(state==2) {
        tft->fillRect(0,20,130,140,BACKGROUND);
        tft->setCursor(0,20,FONTSIZE);
        tft->print("Please Wait");
        sprintf(request,"POST http://608dev.net/sandbox/sc/haoyangg/final_project/friends.py HTTP/1.1\r\n");
        strcat(request,"Host: 608dev.net\r\n");
        strcat(request,"Content-Type: application/x-www-form-urlencoded\r\n");
        sprintf(request+strlen(request),"Content-Length: %d\r\n", strlen(body)); //append string formatted to end of request buffer
        strcat(request,"\r\n"); //new line from header to body
        strcat(request,body); //body
        strcat(request,"\r\n"); //header
        do_http_request("608dev.net", request, response, 1000, 6000,true);
        tft->fillRect(0,20,130,140,BACKGROUND);
        tft->setCursor(0,20,FONTSIZE);
        if(response[0]!='\0') {
          tft->println(response);
        }
        else {
          tft->println("Failed to connect server! Please check your wifi connection");
        }
        tft->println("Long Press to add another friend");
        state=0;
      }
    }
  }
  else {
    if((millis()-check_requests_timer>8000||millis()<check_requests_timer)&&(!do_not_request)) {
      //Serial.println("Add: I will check requests!");
      sprintf(request,"GET http://608dev.net/sandbox/sc/haoyangg/final_project/friends.py?module=add&user=%s&space=%d HTTP/1.1\r\n",user,10-number_of_requests);
      strcat(request,"Host: 608dev.net\r\n");
      strcat(request,"\r\n");
      do_http_request("608dev.net", request, response, 1000, 6000,true);
      char* p=strtok(response,",");
      while(p!=NULL) {
        sprintf(requests_buffer[(start_of_requests+number_of_requests)%10],p);
        number_of_requests++;
        p=strtok(NULL,",");
      }
      check_requests_timer=millis();
    }
  }
}

Nearby_Module::Nearby_Module(TFT_eSPI* tft_to_use,TinyGPSPlus* gps_to_use,char* username) {
  tft=tft_to_use;
  gps=gps_to_use;
  is_active=false;
  user=username;
  timer=millis()-9000;  
}

void Nearby_Module::activate() {
  is_active=true;
  tft->fillRect(0,20,130,140,BACKGROUND);
  tft->setCursor(0,20,FONTSIZE);
  tft->print("Short Press to Refresh");
}

void Nearby_Module::close() {
  is_active=false;
}
void Nearby_Module::update(int bv) {
  if((millis()-timer>9000||millis()<timer)&&(!do_not_request)){
    if(gps->location.isValid()) {
      //Serial.println("Nearby: I will send my lat and lon:");
      //Serial.println(gps->lat);
      //Serial.println(gps->lon);
      sprintf(body,"module=nearby&user=%s&lat=%.6lf&lon=%.6lf",user,gps->location.lat(),gps->location.lng());
      sprintf(request,"POST http://608dev.net/sandbox/sc/haoyangg/final_project/friends.py HTTP/1.1\r\n");
      strcat(request,"Host: 608dev.net\r\n");
      strcat(request,"Content-Type: application/x-www-form-urlencoded\r\n");
      sprintf(request+strlen(request),"Content-Length: %d\r\n", strlen(body)); //append string formatted to end of request buffer
      strcat(request,"\r\n"); //new line from header to body
      strcat(request,body); //body
      strcat(request,"\r\n"); //header
      do_http_request("608dev.net", request, response, 1000, 6000,true);
    }
    timer=millis();
  }
  if(is_active){
    if(bv==1){
      //Serial.println("I will refresh nearby list");
      tft->fillRect(0,20,130,140,BACKGROUND);
      tft->setCursor(0,20,FONTSIZE);
      tft->print("Please Wait");
      sprintf(request,"GET http://608dev.net/sandbox/sc/haoyangg/final_project/friends.py?module=nearby&user=%s HTTP/1.1\r\n",user);
      strcat(request,"Host: 608dev.net\r\n");
      strcat(request,"\r\n");
      do_http_request("608dev.net", request, response, 1000, 6000,true);
      tft->fillRect(0,20,130,140,BACKGROUND);
      tft->setCursor(0,20,FONTSIZE);
      if(response[0]!='\0') {
      tft->println(response);
      }
      else {
        tft->println("Failed to connect server! Please check your wifi connection");
      }
    }
  }
}

Friends_Module::Friends_Module(TFT_eSPI* tft_to_use, TinyGPSPlus* gps_to_use, MPU9255* imu_to_use, char* username):message_module(tft_to_use,imu_to_use,username),add_module(tft_to_use,imu_to_use,username),nearby_module(tft_to_use,gps_to_use,username){
  tft=tft_to_use;
  state=0;
}

void Friends_Module::activate(int a){
  state=a;
  if(a==1) {
    message_module.activate();
  }
  if(a==3) {
    nearby_module.activate();
  }
  if(a==4) {
    add_module.activate();
  }  
}

void Friends_Module::close() {
  if(state==1) {
    message_module.close();
  }
  if(state==3) {
    nearby_module.close();
  }
  if(state==4) {
    add_module.close();
  }
  last_wifi_state=true;
  state=0;
}

void Friends_Module::update(int bv, bool is_wifi_connected) {
  if(state!=0) {
    if(!is_wifi_connected&&last_wifi_state) {
      tft->fillRect(0,20,130,140,TFT_BLACK);
      tft->setCursor(0, 20, 1);
      tft->println("This app requires wifi connection!");
    }
    else if(is_wifi_connected&&!last_wifi_state) {
      activate(state);
    }
    else if(is_wifi_connected) {
      message_module.update(bv);
      nearby_module.update(bv);
      add_module.update(bv);
    }
    last_wifi_state=is_wifi_connected;
  }
  else {
    if(is_wifi_connected) {
      message_module.update(bv);
      nearby_module.update(bv);
      add_module.update(bv);
    }
  }
}

File: Server/friends.py
import sqlite3
import datetime
visits_db = '__HOME__/final_project/friends.db'

def request_handler(request):
    conn = sqlite3.connect(visits_db)
    c = conn.cursor()
    c.execute('''CREATE TABLE IF NOT EXISTS friends (user1 text,user2 text);''')
    c.execute('''CREATE TABLE IF NOT EXISTS location (user text,lat float,lon float,timing timestamp);''')
    c.execute('''CREATE TABLE IF NOT EXISTS messages (user1 text,user2 text,content text,flag int,timing timestamp);''')
    c.execute('''CREATE TABLE IF NOT EXISTS requests (user1 text,user2 text,flag int,timing timestamp);''')
    if request['method']=='POST':
        module=request['form']['module']
        if module=='message':
            user1=request['form']['user1']
            user2=request['form']['user2']
            content=request['form']['content']
            things = c.execute('''SELECT * FROM friends WHERE user1 = ? AND user2 = ?;''',(user1,user2)).fetchall()
            if len(things)==0:
                conn.commit() # commit commands
                conn.close() # close connection to database
                return user2+' is not your friend!'
            else:
                c.execute('''INSERT into messages VALUES (?,?,?,?,?);''', (user1,user2,content,0,datetime.datetime.now()))
                conn.commit() # commit commands
                conn.close() # close connection to database
                return 'Succeed!'
        if module=='add':
            action=request['form']['action']
            if action=='accept':
                user1=request['form']['user1']
                user2=request['form']['user2']
                c.execute('''INSERT into friends VALUES (?,?);''', (user1,user2))
                c.execute('''INSERT into friends VALUES (?,?);''', (user2,user1))
                c.execute('''INSERT into messages VALUES (?,?,?,?,?);''', ('system',user2,user1+' accepted your friend request!',0,datetime.datetime.now()))
                conn.commit() # commit commands
                conn.close() # close connection to database
                return 'Accepted!'
            if action=='refuse':
                user1=request['form']['user1']
                user2=request['form']['user2']
                c.execute('''INSERT into messages VALUES (?,?,?,?,?);''', ('system',user2,user1+' refused your friend request!',0,datetime.datetime.now()))
                conn.commit() # commit commands
                conn.close() # close connection to database
                return 'Refused!'
            if action=='send':
                user1=request['form']['user1']
                user2=request['form']['user2']
                things = c.execute('''SELECT * FROM friends WHERE user1 = ? AND user2 = ?;''',(user1,user2)).fetchall()
                if len(things)!=0:
                    conn.commit() # commit commands
                    conn.close() # close connection to database
                    return user2+' is already your friend!'
                else:
                    things = c.execute('''SELECT * FROM requests WHERE user1 = ? AND user2 = ? AND flag = ?;''',(user1,user2,0)).fetchall()
                    if len(things)!=0:
                        conn.commit() # commit commands
                        conn.close() # close connection to database
                        return 'Repeat sending request!'
                    else:
                        c.execute('''INSERT into requests VALUES (?,?,?,?);''', (user1,user2,0,datetime.datetime.now()))
                        conn.commit() # commit commands
                        conn.close() # close connection to database
                        return "Succeed!"
        if module=='nearby':
            user=request['form']['user']
            lat=request['form']['lat']
            lon=request['form']['lon']
            lat=float(lat)
            lon=float(lon)
            c.execute('''INSERT into location VALUES (?,?,?,?);''', (user,lat,lon,datetime.datetime.now()))
            conn.commit() # commit commands
            conn.close() # close connection to database
            return 'Position!'
    else:
        module=request['values']['module']
        if module=='message':
            user=request['values']['user']
            space=request['values']['space']
            length=int(space)
            things = c.execute('''SELECT * FROM messages WHERE user2 = ? AND flag = ? ORDER BY timing ASC;''',(user,0)).fetchall()
            if len(things)<length:
                length=len(things)
            out=''
            if length!=0:
                for i in range(length-1):
                    (user1,user2,content,flag,timing)=things[i]
                    out+='From '
                    out+=user1
                    out+=':\n'
                    out+=content
                    out+=','
                    c.execute('''UPDATE messages SET flag = ? WHERE user1 = ? AND user2 = ? AND content = ? AND timing = ?;''', (1,user1,user2,content,timing))
                (user1,user2,content,flag,timing)=things[length-1]
                out+='From '
                out+=user1
                out+=':\n'
                out+=content
                c.execute('''UPDATE messages SET flag = ? WHERE user1 = ? AND user2 = ? AND content = ? AND timing = ?;''', (1,user1,user2,content,timing))
            conn.commit() # commit commands
            conn.close() # close connection to database
            return out
        if module=='add':
            user=request['values']['user']
            space=request['values']['space']
            length=int(space)
            things = c.execute('''SELECT * FROM requests WHERE user2 = ? AND flag = ? ORDER BY timing ASC;''',(user,0)).fetchall()
            if len(things)<length:
                length=len(things)
            out=''
            if length!=0:
                for i in range(length-1):
                    (user1,user2,flag,timing)=things[i]
                    out+=user1
                    out+=','
                    c.execute('''UPDATE requests SET flag = ? WHERE user1 = ? AND user2 = ? AND timing = ?;''', (1,user1,user2,timing))
                (user1,user2,flag,timing)=things[length-1]
                out+=user1
                c.execute('''UPDATE requests SET flag = ? WHERE user1 = ? AND user2 = ? AND timing = ?;''', (1,user1,user2,timing))
            conn.commit() # commit commands
            conn.close() # close connection to database
            return out
        if module=='nearby':
            user=request['values']['user']
            twenty_s_ago = datetime.datetime.now()- datetime.timedelta(seconds = 20)
            things = c.execute('''SELECT * FROM location WHERE user = ? AND timing > ? ORDER BY timing DESC;''',(user,twenty_s_ago)).fetchall()
            if len(things)==0:
                conn.commit() # commit commands
                conn.close() # close connection to database
                return 'Please check your GPS!'
            else:
                (me,mylat,mylon,timing)=things[0]
                things = c.execute('''SELECT * FROM friends WHERE user1 = ?;''',(user,)).fetchall()
                out='Nearby Friends (200m):\n'
                number=0
                for x in things:
                    (me,friend)=x
                    things = c.execute('''SELECT * FROM location WHERE user = ? AND timing > ? ORDER BY timing DESC;''',(friend,twenty_s_ago)).fetchall()
                    if len(things)!=0:
                        (friend,lat,lon,timing)=things[0]
                        if ((mylat-lat)**2+(mylon-lon)**2)**0.5<0.002:
                            number+=1
                            out+=friend
                            out+='\n'
                if number==0:
                    conn.commit() # commit commands
                    conn.close() # close connection to database
                    return 'No friend is nearby'
                else:
                    conn.commit() # commit commands
                    conn.close() # close connection to database
                    return out
                    


            
    